File: Green.Core.Tests/CompilerTests.cs
using System;
using System.Collections.Generic;
using System.Text;
using Xunit;

namespace Green.Tests
{
    public class CompilerTests
    {
        [Fact]
        public void Compile_Constant()
        {
            var compiler = new Compiler();
            var expr = new SyntaxConstant(null, 5L);
            var result = compiler.Compile(expr);

            Assert.Equal(new byte[] { (byte)OpCode.Const1, 0, }, result.Code);
            Assert.Equal(new object[] { 5L }, result.Constants);
        }

        [Fact]
        public void Compile_Variable()
        {
            var compiler = new Compiler();
            var expr = new SyntaxIdentifier(null, IdentifierType.Identifier, "a");
            var result = compiler.Compile(expr);

            Assert.Equal(new byte[] { (byte)OpCode.Var1, 0, }, result.Code);
            Assert.Equal(new object[] { "a" }, result.Variables);
        }
    }
}

File: Green.Core.Tests/CompilerTests.fs
module CompilerTests

open Xunit
open Green

[<Fact>]
let Compile_Constant() =
    let expr = SyntaxConstant(null, 5L)
    let result = Compile.Compile expr

    Assert.Equal([ byte OpCode.Const1 ; byte 0 ], result.Code)
    Assert.Equal<obj>([ 5L ], result.Constants)

[<Fact>]
let Compile_Variable() =
    let expr = SyntaxIdentifier(null, IdentifierType.Identifier, "a")
    let result = Compile.Compile expr

    Assert.Equal([ byte OpCode.Var1; byte 0 ], result.Code)
    Assert.Equal([ "a" ], result.Variables)

[<Fact>]
let Compile_Call() =
    let expr = SyntaxList(null,
                          [
                              SyntaxIdentifier(null, IdentifierType.Identifier, "+");
                              SyntaxConstant(null, 1);
                              SyntaxConstant(null, 2);
                          ])
    let result = Compile.Compile expr

    Assert.Equal([
                     byte OpCode.Var1; byte 0;
                     byte OpCode.Const1; byte 0;
                     byte OpCode.Const1; byte 1;
                     byte OpCode.Call1; byte 2;
                 ],
                 result.Code)
    Assert.Equal<obj>([ 1; 2 ], result.Constants)
    Assert.Equal([ "+" ], result.Variables)
